### DISCUSSION TITLE
In the new release index_i index_j flipped in the message_passing layer code

### DISCUSSION THREAD

**Discussion by abmukho** [danielegrattarola/spektral#249] [Q&A]
While comparing my old trained model using crystal conv after implementing the new changes from the new release, I noticed that the index_i and index_j are flipped in the new version in the message_passing code. For example in the previous release, index_i was a.indices[:, 0] which intuitively made more sense because with get_i (self) and get_j(neighbours), the message passing was performed using convolution done on x_i||x_j||e_ij which you then pool on unique x_i's using deserialize scatter. Is there anything subtle in the new release that I am missing? Few explanations that I came up with are that either e_ij is now actually e_ji or a_ij is now a_ji... which I didn't see in the code. What am I missing? This will be a problem in e.g. bidirectional asymmetric edge matrix.

**Comment by danielegrattarola** [comment]
Hi, 

the message passing should be from source (neighbors) to targets, so get_j() should actually return the a.indices[:, 0] because that's the "source" side of the edges.
The order with which we gather edges did not change with this modification, i.e. the first row of x_i, x_j and e are always associated with the first edge. It's just the flow of information that's reversed. 

But yeah, in your example it should be x_i | x_j | e_ji. 
I'll fix the docs 

Cheers

**Reply by 097475** [reply]
I'm sorry to resurrect such an old conversation, but in the getting started page (https://graphneural.network/getting-started/) it is implied that the edges originate in the j-index and target the i-index. Therefore if in my sparse adjacency matrix I have an entry with coordinates (0, 5) it means an edge from node 5 is going into node 0. But the get_j() method in this case would select 0 as source node. 
So I don't understand which format I should use for my adjacency matrices, should I use (0, 5) or (5, 0) for an edge going from 5 to 0?

**Reply by danielegrattarola** [reply]
I don't know what possessed me to name those methods get_i and get_j, it's just so confusing. I'll fix it in the next release to something more human. 

Anyway, the answer is that the initial implication in your post is incorrect. A non-zero entry at position (0, 5) in the adjacency matrix indicates an edge FROM 0 TO 5. 
So `get_j` correctly selects node 0 as a neighbour/source of the message and `get_i` correctly selects node 5 as the root/target.

Can you point out from where in the documentation you got that initial assumption? So I can fix it.

Thanks

**Reply by 097475** [reply]
Thank you for your response.

From the getting started page there is this sentence:

> Each entry a[i, j] of the adjacency matrix is non-zero if there exists an edge going from node j to node i, and zero otherwise.

And this example:

R, C, V
0, 1, 1
0, 3, 2
1, 0, 3
1, 2, 4
2, 1, 5
2, 3, 6
3, 0, 7
3, 2, 8

> where R indicates the "row" indices, C the columns, and V the non-zero values a[i, j]. For example, in the second line, we see that there is an edge that goes from node 3 to node 0 with weight 2.

**Reply by danielegrattarola** [reply]
Yeah those are mistakes, I must have mixed things up when writing the example.

The R column is "from", the C column is "to", and `a[i, j]` gives you the edge from i to j. Does that make sense?
I'll fix it, thanks for noticing.

**Reply by 097475** [reply]
Yes, that absolutely makes sense. Thank you very much for your time!
